### ISSUE TITLE
the application doesn't crawl all links on file

### ISSUE THREAD

**Issue by AhmadAzmiDaoud** [HTTPSChecker/releases#27]
Hi,
i purchase enterprise plan and i want to scan text file -find the attached-, but the tool stopped after 19 link. could you help on that, find the attached picture
[6.txt](https://github.com/HTTPSChecker/releases/files/2155904/6.txt)
![pic](https://user-images.githubusercontent.com/40769143/42174236-e7f82228-7e21-11e8-9314-e81afd667f3e.jpg)

**Comment by stilliard** [comment]
Hi Ahmad,
Thank you for contacting us. 

I've had a look over the file supplied and found the reason it stops after the 19th url is that our crawls can only run on one origin / subdomain at a time, the first 19 urls run on www.aljazeera.net but then the 20th and 21st run on blogs.aljazeera.net and the following run on midan.aljazeera.net. 

The different subdomains will need to be run as separate crawls.

I hope this is ok, please let me know if you have any further questions, thank you.

**Comment by AhmadAzmiDaoud** [comment]
Thank you very much...
We have another issue, the following attached file have 50,000 all of them related to the same domain, when we try to scan them, but no pages scanned , could you help on that
![untitled](https://user-images.githubusercontent.com/40769143/42211493-0589b422-7eb4-11e8-9192-2cfef7420dca.png)
[Book4.txt](https://github.com/HTTPSChecker/releases/files/2158593/Book4.txt)

**Comment by stilliard** [comment]
You're welcome, happy to help out.

I've taken a look at the new file but it seems the first url (and maybe the others) don't load for me.

Are these URLs only accessible within a firewall or proxy maybe?

E.g. if I request https://www.aljazeera.net/home/getpage/97BF4C0E-B160-4E00-8A8B-AF44BE2D88AE/EB2185BD-EAD6-4404-A0E0-FB28B47AB546  with curl or in my browser it doesn't load & eventually times out. 
```
$ curl -I https://www.aljazeera.net/home/getpage/97BF4C0E-B160-4E00-8A8B-AF44BE2D88AE/EB2185BD-EAD6-4404-A0E0-FB28B47AB546
HTTP/1.1 504 Gateway Time-out
```

To help with this see the crawl errors file, it's available from that report screen by clicking the Crawl Logs button.

I hope this helps.

**Comment by AhmadAzmiDaoud** [comment]
Hi,
yes we use with a new environment that just accessible from my side, but when i try to scan the URL direct it is working fine, see the attached please....

![image](https://user-images.githubusercontent.com/40769143/42213335-c09aa114-7eb8-11e8-8b68-1f29cb1a82c7.png)

![image](https://user-images.githubusercontent.com/40769143/42213422-f7df8996-7eb8-11e8-83d0-ccc82f053226.png)

**Comment by stilliard** [comment]
Ah ok, no problem.

I think the best place to check then is in the logs, I've just noticed the logs button doesn't show atm if the scan found no issues (I'll get this changed soon so the logs button is always there). 

But you can find the logs folder by opening up the following folder on your machine: 
%USERPROFILE%\AppData\Roaming\httpschecker\

There should be an error log in here that should help show any issues it found. 
If you run the scan again using the import file it should create this log file for you.
If the file doesn't seem to be of much help though please send it over to me & I can take a look. 

Thank you.

**Comment by AhmadAzmiDaoud** [comment]
[WWW-1111111.txt](https://github.com/HTTPSChecker/releases/files/2162951/WWW-1111111.txt)
Hi,
i have the following error on the debug log
HTTPS Error: ESOCKETTIMEDOUT

and the error logs files are empty, see the attached...
![untitled](https://user-images.githubusercontent.com/40769143/42273177-1db93440-7f91-11e8-836b-bbaadfa432e6.jpg)

while i can open this article on the browser, see the second attached picture

![image](https://user-images.githubusercontent.com/40769143/42273398-b86eec1e-7f91-11e8-8acc-f6bc15bc84fd.png)

first article link on the file:
http://www.aljazeera.net/home/getpage/030BFE32-A88C-44AA-A993-190BC36F0CB5/00826103-1DC8-4C1B-9F8B-F15DC8E1EDF8

see the current logs please:
HTTPS Error: ESOCKETTIMEDOUT
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/3676C54A-2CA1-4204-A1F5-2A438C1AA476 (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/3685F1E3-6ED6-426B-9410-08F959EF729F (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/368CB0B2-7CDE-4821-9EB8-1B6395961641 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/36FA01E8-CA11-4FC8-9C2D-03252841CF10 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/38461E05-DC9E-41BD-BB08-998AE42C0FF8 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/3F658F8E-C4E7-41CC-825E-740C5276CD55 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/42B9AB63-B133-4CA2-A9EB-34F2B20A8AFA (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/42D2C381-236D-4B60-ABF8-A8EE3EF8AAAC (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/4353C058-1E47-4356-BC75-F9877FA6EAB0 (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/43C4AAC5-BDE5-4C54-AC22-2775CFC339F0 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/4406994F-435F-4B16-AD54-5B2EE9C63A55 (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/4F622951-8EE7-4A71-80E3-4C353F56CD2E (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/4F988232-3FF5-4F61-AF7E-2877FE9D7839 (Error: read ECONNRESET)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/50E5F415-E7AB-4F2C-86A2-701B3A2E2551 (Error: socket hang up)
Fetch Client Error: Request died locally for https://www.aljazeera.net/home/getpage/0353E88A-286D-4266-82C6-6094179EA26D/5397B49B-1958-4CE6-A25A-170C224D12D5 (Error: read ECONNRESET)

**Comment by stilliard** [comment]
Hi Ahmad, 

Thanks for sending over the debug log details. 

From the errors it looks like the issues are caused by the requests not connecting & timing out.

When I try to load these pages too I get the same though I know you mentioned it's behind a firewall. 

My only thought is I wonder if you need to allow our app to connect into this area too. 
It'll be making requests direct from your machine but maybe it needs to be allowed through?

**Comment by AhmadAzmiDaoud** [comment]
Hi,
Yes there is firewall, but the tool installed in a server that can access the domain, you can see the snapshot on the below link that i sent before where the chrome browser opened the page successfully but the Application raise the same error

https://user-images.githubusercontent.com/40769143/42273398-b86eec1e-7f91-11e8-8acc-f6bc15bc84fd.png

page link:
http://www.aljazeera.net/home/getpage/030BFE32-A88C-44AA-A993-190BC36F0CB5/00826103-1DC8-4C1B-9F8B-F15DC8E1EDF8

https://www.aljazeera.net/home/getpage/030BFE32-A88C-44AA-A993-190BC36F0CB5/00826103-1DC8-4C1B-9F8B-F15DC8E1EDF8

do the application use specific port or something ? we just allow the port 443 for our site, do you need ping to be opened or anything else?

Notes could help:
-the tool working fine on the document contain 19 links.
-all URLs on the big document make 301 redirect to the friendly URL

i can connect with u on team viewer if this may help you
thanks

**Comment by stilliard** [comment]
Hi Ahmad,

That is strange, our app doesn't use ping or anything, just straight forward requests over port 80 & 443 like Chrome.

My only other ideas are that either:
1. Are you using a proxy setup within Chrome to connect through the firewall?

2. Maybe the firewall on the server is checking more than just your IP? Maybe it's also looking at your User-agent to check if it's really Chrome? You can change to use Windows Chrome User-agent in the advanced settings to try to get around this.

or 3. Does crawling any website work through the app for you? (if not it might be you have a firewall on your local machine blocking the app making requests).

**Comment by AhmadAzmiDaoud** [comment]
Hi,
1.No.
2.i tried that with no luck.
3.it works for the same website (aljazeera.net) when i put the URL direct -not choosing a file-

also another problem: when just i put the site name and scan, and i tried to stop the scan but the application not respond - see the second pic attached
i can give an access throw team viewer to check the case

![image](https://user-images.githubusercontent.com/40769143/42304212-b73eb0bc-802d-11e8-9e37-028fc38aecb6.png)

![image](https://user-images.githubusercontent.com/40769143/42304358-7aaf99da-802e-11e8-96cd-c0a520ab0fb5.png)

**Comment by stilliard** [comment]
Good Morning Ahmad,
Apologies this issue is still on going. 

As I believe the issue is due to the Firewall I believe you'll need to speak to whoever is in control of this. 

But I do have an idea that might help yourself & them with finding out what's different between our requests & Chrome.

You could install a free debugging proxy onto your machine such as Charles: https://www.charlesproxy.com/download/ 

Once it's running, open HTTPS Checker and inside Advanced Options > Proxy > Proxy URL enter http://localhost:8888 

Then start your scan as normal and it should log every request it as it makes them. 
You'll be able to see everything about how we make the requests in here which should be able to help the person running the firewall to see what's causing the issue and adapt the firewall to allow the app through or let us know what we can do to get our app through it as we'd be happy to help out too. 

If you're not able to get in touch with who's running the firewall, another option could be to use the developer tools in chrome to see the Network request for a page that's not loading in our app. It should reveal the headers sent etc and you could compare these with our app's ones shown in the Charles proxy. If you find you need to send an extra couple HTTP headers with the request you can do so through Advanced Options > General > Custom HTTP Headers.

I hope this helps & I hope the person running the firewall is able to help more to get the app allowed through.

**Comment by AhmadAzmiDaoud** [comment]
Hi,
now we published the site over CDN and the site is public now under this domain: 
https://ww-ak.aljazeera.net/portal

and we still have the same issue when we upload file to be scanned, find the attached file please...
[Book4.txt](https://github.com/HTTPSChecker/releases/files/2210926/Book4.txt)
![image](https://user-images.githubusercontent.com/40769143/42958403-1b0b2362-8b8e-11e8-9517-647582a3e3fe.png)

**Comment by stilliard** [comment]
Hi Ahmad,
Thanks for linking to the public url so we can test. 
I've run some demo scans on this & I think I can now see the issue. 

Currently all 50,000 urls in the list redirect, which is fine as we find any redirects and then re-queue the redirect url to be crawled. But as we have 50,000 urls pushed into the queue in front of them, each redirect is behind the existing requests & so the crawler must first go through every one of those 50,000 requests first, showing no increase in page counts while it does to then find the redirect url. 
Then i've found that the pages they redirect to all redirect again after leading to them being another 50,000 behind the queue.

This means that the scan of this file wont properly start until it's made the first 100,000 redirect requests which will appear very slow. 

To test this I cut the file down to 5 urls instead and it was able to crawl these without issue. 

I have faith our app will have no problem crawling these links, but as the scan wont actually find a url until it's done all those redirects it may be hitting what we call the "Crawl Bail Time". This is a timeout we have in the app to give up on the scan if it appears it's taking too long without any new pages found. 

I think if you set the Crawl Bail Time to 86400 seconds (which works out as 1 day which is longer than it should need but feels like a good safe point to give up at) in Advanced Options > Limits, this should let the app crawl the site, though it will appear that it's doing nothing while it crawls those first 100,000 redirects which may hours to complete.  

How does this sound to you?

**Comment by AhmadAzmiDaoud** [comment]
sounds good, unfortunately my subscription end today before we start using your app due the problems we have with the application , and now i cant test this solution, could you extend my subscription that related to: Hostingaccounts@ihorizons.com

**Comment by stilliard** [comment]
Hi Ahmad, 
I've checked your subscription which should last until the 26th.
However as you've had some issues using the app from behind the firewall before, if you let the subscription auto renew, then once it has and on this occasion as a gesture of goodwill, we will refund we will refund your first months cost for you as you haven't had much time to use it. 
How does that sound to you?

**Comment by AhmadAzmiDaoud** [comment]
Dear Stilliard , 
kindly note that we still have the auto renew on but we only need the tool for a single month and since we didn't make a use of it for the previous month due to the bulk URLs issue will we get a refund for the previous month? or shall we disable the auto renew and you will activate the license for another month? 

Regards,

**Comment by stilliard** [comment]
Hi Ahmad,
Sure no problem at all. 

We use Gumroad to manage our subscriptions but they don't currently offer a way for us to extend a licence. However to get around this is you keep the auto renew on so it charges for another month, we'll refund you your first month. This way you have only paid for the one month but you actually get the subscription for this next month as well.

I hope this helps.

**Comment by AhmadAzmiDaoud** [comment]
Thank you very much for your continuous support...

I have one last problem, i cant get the links that raised errors when the tool finished scanning, even from the logs file in %USERPROFILE%\AppData\Roaming\httpschecker\

To explain:
For example; when i give the tool file contains 5000 URLs, it scan 4800 URLs and skipped 200 URLs , while the tool running i can see the URLs that raises errors  when i press the debug log button. such as timeout error, but when the tool finished scanning i cant get any of them even from the logs file.

**Comment by stilliard** [comment]
Hi @AhmadAzmiDaoud ,
Very sorry for my late reply, not sure how I missed your latest email. 

All the urls that error should be inside the crawl errors log file in that folder, but if they are still not appearing for you could you send me over the file used in the scan and the log files you received. 

Thank you.
